src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousType.TypePublicSymbol.cs

```diff
@@ -46,41 +40,26 @@ internal AnonymousTypePublicSymbol(AnonymousTypeManager manager, AnonymousTypeDe
                 this.Manager = manager;
                 this.TypeDescriptor = typeDescr;
 
-                int fieldsCount = typeDescr.Fields.Length;
+                var fields = typeDescr.Fields;
+                var properties = fields.SelectAsArray((field, i, type) => new AnonymousTypePropertySymbol(type, field, i), this);
 
                 //  members
-                Symbol[] members = new Symbol[fieldsCount * 2 + 1];
-                int memberIndex = 0;
+                int membersCount = fields.Length * 2 + 1;
+                var members = ArrayBuilder<Symbol>.GetInstance(membersCount);
 
-                // The array storing property symbols to be used in 
-                // generation of constructor and other methods
-                if (fieldsCount > 0)
+                foreach (var property in properties)
                 {
-                    AnonymousTypePropertySymbol[] propertiesArray = new AnonymousTypePropertySymbol[fieldsCount];
-
-                    // Process fields
-                    for (int fieldIndex = 0; fieldIndex < fieldsCount; fieldIndex++)
-                    {
-                        // Add a property
-                        AnonymousTypePropertySymbol property = new AnonymousTypePropertySymbol(this, typeDescr.Fields[fieldIndex]);
-                        propertiesArray[fieldIndex] = property;
-
-                        // Property related symbols
-                        members[memberIndex++] = property;
-                        members[memberIndex++] = property.GetMethod;
-                    }
-
-                    this.Properties = propertiesArray.AsImmutableOrNull();
-                }
-                else
-                {
-                    this.Properties = ImmutableArray<AnonymousTypePropertySymbol>.Empty;
+                    // Property related symbols
+                    members.Add(property);
+                    members.Add(property.GetMethod);
                 }
 
+                this.Properties = properties;
+
                 // Add a constructor
-                members[memberIndex++] = new AnonymousTypeConstructorSymbol(this, this.Properties);
-                _members = members.AsImmutableOrNull();
-                Debug.Assert(memberIndex == _members.Length);
+                members.Add(new AnonymousTypeConstructorSymbol(this, properties));
+                _members = members.ToImmutableAndFree();
```

Comment:
>ToImmutableAndFree [](start = 35, length = 18)

I don't know if it matters, but do `.AsImmutableAsNull()` and `.ToImmutableAndFree()` differ in empty case?

Reply:
Both should return `Empty` in that case.

---
In reply to: [259071168](https://github.com/dotnet/roslyn/pull/33562#discussion_r259071168) [](ancestors = 259071168)